### PR TITLE
apollo-encoder: align encoder closer to graphql specs #161

### LINKED ISSUE #161: apollo-encoder: align encoder closer to graphql spec
## Description

Currently `apollo-encoder` supports creation of any GraphQL Type System definitions. It was primarily written to allow Rover to encode introspection response into a valid SDL. To ensure that `apollo-encoder` can be used for any definition in the spec, we need to adjust and add a series of definitions. This will especially be useful to support normalisation work that Apollo Router is doing.

## Implementation
The following definitions need to be added:
- [x] Operation Definition
- [x] Fragment Definition
- [x] Type System Extensions
    - [x] Schema Extension
    - [x] Scalar Extension
    - [x] Interface Extension
    - [x] Union Extension
    - [x] Enum Extension
    - [x] Input Object Extension

The following definitions need to be adjusted
- [x] Input Field and Field are two different types in the GraphQL type system and they need to be separated
- [x] Input Value Definitions can accept any type of directive, not just `deprecated`
- [x] Schema Definition can accept any type of directive
- [x] Scalar Definition can accept any type of directive
- [x] Object Definition can accept any type of directive
- [x] Interface Definition can accept any type of directive
- [x] Union Definition can accept any type of directive
- [x] Enum Definition can accept any type of directive
- [x] Input Object Definition can accept any type of directive

### PR DESCRIPTION
**Pull request by bnjjj**
close  #161

### CODE REVIEW

File: crates/apollo-encoder/src/fragment.rs [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+        write!(f, "...{}", self.name)?;
+        for directive in &self.directives {
+            write!(f, " {}", directive)?;
+        }
+
+        Ok(())
+    }
+}
+
+/// The __inlineFragment type represents an inline fragment that could be used as a field
```
Comment (lrlna) [inline crates/apollo-encoder/src/fragment.rs:123]:
```suggestion
/// The __inlineFragment type represents an inline fragment in a selection set that could be used as a field
```

File: crates/apollo-encoder/src/fragment.rs [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+
+    /// Set the inline fragment's type condition.
+    pub fn type_condition(&mut self, type_condition: Option<TypeCondition>) {
+        self.type_condition = type_condition;
+    }
+
+    /// Should be used everywhere in this crate isntead of the Display implementation
+    /// Display implementation is only useful as a public api
+    pub(crate) fn format_with_indent(&self, indent_level: usize) -> String {
+        let mut text = String::new();
```
Comment (lrlna) [inline crates/apollo-encoder/src/fragment.rs:185]:
```suggestion
        let mut text = String::from("...");
```

File: crates/apollo-encoder/src/input_value.rs [hunk 3]
```diff
@@ -1,14 +1,14 @@
 use std::fmt;
 
-use crate::{StringValue, Type_};
+use crate::{Directive, StringValue, Type_};
 
 // NOTE(@lrlna): __InputValue is also meant to be used for InputFields on an
 // InputObject. We currently do not differentiate between InputFields and
 // Fields, so this is not applied directly to InputObjects. Once we are able to
 // walk an AST to encode a schema, we will want to make sure this struct is used
 // directly: InputObject --> InputField --> InputValue
```
Comment (lrlna) [inline crates/apollo-encoder/src/input_value.rs:9]: you've made fixes that make this note irrelevant! 🎉 
```suggestion
```

File: crates/apollo-encoder/src/interface_def.rs [hunk 4]
```diff
@@ -134,7 +158,7 @@ impl fmt::Display for InterfaceDef {
 #[cfg(test)]
 mod tests {
```
Comment (lrlna) [inline crates/apollo-encoder/src/interface_def.rs:159]: Let's also add a test with an extension here.

File: crates/apollo-encoder/src/input_object_def.rs [hunk 5]
```diff
@@ -97,7 +122,7 @@ impl fmt::Display for InputObjectDef {
 #[cfg(test)]
 mod tests {
```
Comment (lrlna) [inline crates/apollo-encoder/src/input_object_def.rs:123]: Let's also add a test with an input object definition extension.

File: crates/apollo-encoder/src/enum_def.rs [hunk 6]
```diff
@@ -84,6 +112,8 @@ impl fmt::Display for EnumDef {
 
 #[cfg(test)]
 mod tests {
+    use crate::{Argument, Value};
```
Comment (lrlna) [inline crates/apollo-encoder/src/enum_def.rs:115]: let's add a test with an enum definition extension

File: crates/apollo-encoder/src/enum_value.rs [hunk 7]
```diff
@@ -1,6 +1,6 @@
 use std::fmt;
 
-use crate::StringValue;
+use crate::{Directive, StringValue};
```
Comment (lrlna) [inline crates/apollo-encoder/src/enum_value.rs:3]: nit: I saw you grouped Field Definition and Field in the same file, we can probably do the same for Enum Definition and Enum Value.

File: crates/apollo-encoder/src/operation.rs [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+                query($variable_def: [Int]) @testDirective(first: "one") {
+                  first
+                  second
+                }
+            "#}
+        );
+    }
+
+    #[test]
+    fn it_encodes_a_deeper_query_operation() {
```
Comment (lrlna) [inline crates/apollo-encoder/src/operation.rs:197]: these are great 🙏 !

File: crates/apollo-encoder/src/operation.rs [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+    variable_definitions: Vec<VariableDef>,
+    directives: Vec<Directive>,
+    selection_set: SelectionSet,
+}
+
+impl OperationDef {
+    /// Create a new instance of OperationDef
+    pub fn new(operation_type: OperationType, selection_set: SelectionSet) -> Self {
+        Self {
+            operation_type,
```
Comment (lrlna) [inline crates/apollo-encoder/src/operation.rs:67]: How do you handle query shorthand encoding that doesn't have the "query" keyword, but is indeed of `Query` operation type?

File: crates/apollo-encoder/src/schema.rs [hunk 10]
```diff
@@ -33,6 +33,7 @@ use crate::{
 /// );
 /// ```
 #[derive(Debug)]
+// TODO delete in profit of Document
```
Comment (lrlna) [inline crates/apollo-encoder/src/schema.rs:36]:
```suggestion
// TODO delete in favour of Document
```

Comment (lrlna) [reply crates/apollo-encoder/src/schema.rs:36]: And yea, let's do this as a second pass after we merge this PR.

File: crates/apollo-encoder/src/schema_def.rs [hunk 11]
```diff
@@ -127,4 +154,24 @@ mod tests {
         "#}
         );
     }
+
+    #[test]
```
Comment (lrlna) [inline crates/apollo-encoder/src/schema_def.rs:158]: let's also add an extension test

File: crates/apollo-encoder/src/selection_set.rs [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+    /// Add a selection in the SelectionSet
+    pub fn selection(&mut self, selection: Selection) {
+        self.selections.push(selection);
+    }
+
+    /// Should be used everywhere in this crate isntead of the Display implementation
+    /// Display implementation is only useful as a public api
+    pub(crate) fn format_with_indent(&self, mut indent_level: usize) -> String {
+        let mut text = String::new();
+        text.push_str("{\n");
```
Comment (lrlna) [inline crates/apollo-encoder/src/selection_set.rs:60]:
```suggestion
        let mut text = String::from("{\n");
```

File: crates/apollo-encoder/src/selection_set.rs [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+        } else {
+            text.push_str(&format!("{}}}", "  ".repeat(indent_level - 1)));
+        }
+
+        text
+    }
+}
+
+// This impl is only useful when we generate only a SelectionSet
+// If it's used from a root element, we call `format_with_indent`
```
Comment (lrlna) [inline crates/apollo-encoder/src/selection_set.rs:81]: What do you mean by `root` here? The top-level `Document`?

File: crates/apollo-encoder/src/union_def.rs [hunk 14]
```diff
@@ -58,13 +73,24 @@ impl UnionDef {
...
-        write!(f, "union {} = ", self.name)?;
+        write!(f, " =")?;
 
         for (i, member) in self.members.iter().enumerate() {
             match i {
-                0 => write!(f, "{}", member)?,
+                0 => write!(f, " {}", member)?,
                 _ => write!(f, " | {}", member)?,
             }
         }
```
Comment (lrlna) [inline crates/apollo-encoder/src/union_def.rs:96]: let's have an extension test here as well

File: crates/apollo-encoder/src/value.rs [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+}
+
+impl fmt::Display for Value {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Variable(v) => write!(f, "${v}"),
+            Self::Int(i) => write!(f, "{i}"),
+            Self::Float(fl) => write!(f, "{fl}"),
+            Self::String(s) => {
+                if s.contains('"') {
```
Comment (lrlna) [inline crates/apollo-encoder/src/value.rs:38]: it's also a block string character if it contains `\n` or `\r`
```suggestion
                if s.contains('"') | s.contains('\n') | s.contains('\r') {
```

**Review by lrlna — CHANGES_REQUESTED** [review]
This is looking super good! Excited to have this published soon. 

The biggest thing I don't see resolved yet is the using query shorthand for an operation definition. Maybe I missed the code for it?

File: crates/apollo-encoder/src/selection_set.rs (hunk 13 above)
Comment (bnjjj) [reply crates/apollo-encoder/src/selection_set.rs:80]: I should use `parent` instead

File: crates/apollo-encoder/src/enum_value.rs (hunk 7 above)
Comment (bnjjj) [reply crates/apollo-encoder/src/enum_value.rs:3]: We also have `field.rs` and `field_value.rs` I think it's good to keep the same. If you think it's weird then I can also delete the `field_value.rs` and put everything inside. As you wish

File: crates/apollo-encoder/src/operation.rs (hunk 9 above)
Comment (bnjjj) [reply crates/apollo-encoder/src/operation.rs:67]: It was not supported. It's fixed :)

File: crates/apollo-encoder/src/enum_value.rs (hunk 7 above)
Comment (lrlna) [reply crates/apollo-encoder/src/enum_value.rs:3]: ah! i missed the field.rs and field_value.rs differentiation. let's keep it separate then

File: crates/apollo-encoder/src/selection_set.rs (hunk 13 above)
Comment (lrlna) [reply crates/apollo-encoder/src/selection_set.rs:80]: yes `parent` makes more sense to me!

**Comment by bnjjj** [comment]
I made every changes except the one with `enum_value.rs` check my answer on your comment. Also I used this https://github.com/apollographql/apollo-rs/pull/162/files#diff-c5727d7d12a744bf48d8d53af8c6d9aa98316eaa005f7e3fb4e0be813fbcdf66R1 to keep the readme up to date with our documentation and our code.

**Review by lrlna — APPROVED** [review]
woot woot! let's squash and merge it!

**Pull request merged**